plugin/trino-hive/src/main/java/io/trino/plugin/hive/ReaderProjectionsAdapter.java

```diff
@@ -150,41 +150,24 @@ private Block loadInternalBlock(List<Integer> dereferences, Block parentBlock)
 
         private Block adaptNulls(ColumnarRow columnarRow, Block loadedInternalBlock)
         {
+            if (!columnarRow.mayHaveNull()) {
+                return loadedInternalBlock;
+            }
             // TODO: The current implementation copies over data to a new block builder when a null row element is found.
             //  We can optimize this by using a Block implementation that uses a null vector of the parent row block and
             //  the block for the field.
-
-            BlockBuilder newlyCreatedBlock = null;
+            BlockBuilder newlyCreatedBlock = type.createBlockBuilder(null, columnarRow.getPositionCount());
             int fieldBlockPosition = 0;
 
             for (int i = 0; i < columnarRow.getPositionCount(); i++) {
-                boolean isRowNull = columnarRow.isNull(i);
-
-                if (isRowNull) {
-                    // A new block is only created when a null is encountered for the first time.
-                    if (newlyCreatedBlock == null) {
-                        newlyCreatedBlock = type.createBlockBuilder(null, columnarRow.getPositionCount());
-
-                        // Copy over all elements encountered so far to the new block
-                        for (int j = 0; j < i; j++) {
-                            type.appendTo(loadedInternalBlock, j, newlyCreatedBlock);
-                        }
-                    }
+                if (columnarRow.isNull(i)) {
                     newlyCreatedBlock.appendNull();
                 }
                 else {
-                    if (newlyCreatedBlock != null) {
-                        type.appendTo(loadedInternalBlock, fieldBlockPosition, newlyCreatedBlock);
-                    }
-                    fieldBlockPosition++;
+                    type.appendTo(loadedInternalBlock, fieldBlockPosition++, newlyCreatedBlock);
                 }
             }
```

Comment:
Isn't it still valuable to do this? `mayHaveNull` can return `true` and the block could still contain no nulls -- after all, the only definitive answer is when it returns `false`.

Reply:
That's fair. I was presuming that the changes in `RowBlock` to detect no nulls present at creation time were now going to handle that scenario and would be expected to return `false` from `mayHaveNull` unless there _was_ at least one null row, but I suppose you could find a way to do a range based `RowBlock#getRegion` somewhere in between the two and end up doing unnecessary work in my revised formulation. I've added back an outer probe loop and an inner loop for once a null position is found.